File: myFsharpProject/ProductServices.fs
namespace myFsharpProject
module Product =

    open myFsharpProject.BusinessObjects
    open myFsharpProject.EncryptionComponent

    type Services() =

        member this.GetOne(id: int) =
            let fromDb = Async.RunSynchronously ( Database.getById id )
            fromDb

        member this.GetMany(ids: int[]) =
            let mapped = Array.map (fun id -> Database.getById id) ids
            let result = mapped |> Async.Parallel |> Async.RunSynchronously
            result

        member this.Create(info: CreateNew) =
            let encrypted = encryptFormula info.Formula
            let shared = { SharedProduct.Id = info.Id; Name = info.Name; Formula = Async.RunSynchronously(encrypted) }
            shared

        member this.Create(info: CreateFrom) =
            let formulas =  info.Existing |> Array.map (fun e -> e.Formula)  |> decryptParallel |> Array.concat

            let newFormula = Array.append formulas info.Formula
            let encrypted = encryptFormula newFormula
            let shared = { SharedProduct.Id = info.Id; Name = info.Name; Formula = Async.RunSynchronously encrypted }
            shared

        member this.Query(info: QueryInfo) =
            let { Predicate = predicate; Formula = formula } = info
            let decrypted = decryptFormula formula |> Async.RunSynchronously
            let found = Array.filter predicate decrypted
            let result =  found.Length > 0
            result

File: ConsoleProject/ProductServices.cs
using System;
using System.Collections.Generic;
using Microsoft.FSharp.Core;
using myFsharpProject;

namespace ConsoleProject
{
	public class ProductServices
	{
		/// Strings in C# are inmutable!!!
		/// If we re-write function to encrypt/decrypt in Portal
		/// in a F# func that takes as input string (xml) and
		/// returns string (encrypted xml) we could make it faster! 
		public static void Test()
		{
			var service = new Services();

			// Create
			Component[] components = {
				new Component("Agua", 10.9),
				new Component("Cromo", 10.9),
				new Component("Manganeso", 10.9)
			};
			CreateNew info = new CreateNew(100, "Tumbirichato de amonio", components);
			SharedProduct product = service.Create(info);

			Console.WriteLine("{0}", product);

			// Merge
			var prodArray = service.GetMany(new int[] { 1, 2, 3, 4, 5, 6 });

			CreateFrom fromInfo = new CreateFrom(101, "Kit Product", components, prodArray);
			SharedProduct product2 = service.Create(fromInfo);

			Console.WriteLine("{0}", product2);

			//var predicate = FuncConvert.ToFSharpFunc((Component c) => c.Name == "Poison" && c.Percent > 10);
			//QueryInfo queryInfo = new QueryInfo(predicate, product2.Formula);
			//bool result = service.Query(queryInfo);

			//Console.WriteLine("product2 has Poison > 10% = {0}", result);
		}
	}
}
